staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -0,0 +1,189 @@
+	clone.Header.Del(wsstream.WebSocketProtocolHeader)
+	clone.Header.Del("Sec-Websocket-Key")
+	clone.Header.Del("Sec-Websocket-Version")
+	clone.Header.Del(httpstream.HeaderUpgrade)
+	// Update the http request for an upstream SPDY upgrade.
+	clone.Method = "POST"
+	clone.Body = nil // Remove the request body which is unused.
+	clone.Header.Add(httpstream.HeaderUpgrade, spdy.HeaderSpdy31)
+	clone.Header.Add(httpstream.HeaderProtocolVersion, protocol)
+	return clone
+}
+
+var _ http.ResponseWriter = &tunnelingWriter{}
+var _ http.Hijacker = &tunnelingWriter{}
+
+// tunnelingWriter implements the http.ResponseWriter and http.Hijcker
+// interfaces. After "Hijack", the stored ResponseWriter will not work.
+type tunnelingWriter struct {
+	conn net.Conn
+	w    http.ResponseWriter
+}
+
+func createTunnelingResponseWriter(w http.ResponseWriter, tunnel net.Conn) http.ResponseWriter {
+	return &tunnelingWriter{conn: tunnel, w: w}
+}
+
+// Hijack returns a "net.Conn" which tunnels SPDY through WebSockets.
+// The returned bufio.ReadWriter and error are always nil.
+func (w *tunnelingWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
+	klog.V(6).Infof("Hijack returning websocket tunneling net.Conn")
+	return w.conn, nil, nil
+}
+
+// Header is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingWriter) Header() http.Header {
+	return w.w.Header()
+}
+
+// Write is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingWriter) Write(p []byte) (int, error) {
+	return w.w.Write(p)
+}
+
+// WriteHeader is delegated to the stored "http.ResponseWriter".
+func (w *tunnelingWriter) WriteHeader(statusCode int) {
+	w.w.WriteHeader(statusCode)
+}
+
+// headerInterceptingConnection wraps the tunneling "net.Conn" to drain the
+// HTTP response from the upstream SPDY connection, since the client has
+// already received a "101 Switching Protocols" from the WebSockets
+// upgrade.
+type headerInterceptingConnection struct {
+	net.Conn
+
+	headerBuffer     *bytes.Buffer
+	completedHeaders bool
+	parsedStatus     string
+	parsedHeaders    http.Header
+}
+
+// Write intercepts to initially swallow the HTTP response, then
+// delegate to the tunneling "net.Conn" once the response has been
+// seen and processed.
+func (h *headerInterceptingConnection) Write(b []byte) (int, error) {
+	if h.completedHeaders {
+		return h.Conn.Write(b)
+	}
+
+	// Write into the headerBuffer, then attempt to parse the bytes
+	// as an http response.
+	n, err := h.headerBuffer.Write(b)
+	if err != nil {
+		return n, err
+	}
+	bufferedReader := bufio.NewReader(h.headerBuffer)
+	resp, err := http.ReadResponse(bufferedReader, nil)
+	if errors.Is(err, io.ErrUnexpectedEOF) {
+		// don't yet have a complete set of headers
+		// TODO: reset headerBuffer to append to end and read from start
+		return n, nil
+	}
+	if err != nil {
+		klog.Errorf("invalid headers: %v", err)
+		return n, err
+	}
+
+	h.completedHeaders = true
+	h.parsedStatus = resp.Status
```

Comment:
should we check the resp.Status?

Reply:
definitely, looking at the headers and status that come back to make sure it is an upgrade and make sure the backend selected a spdy protocol is on the list

Reply:
I actually need to do the latter as part of shifting the websocket upgrade back to the client after the backend has indicated it can upgrade

Reply:
The `UpgradeAware` proxy is also handling/validating the SPDY response. We should never get to the `headerInterceptingConnection` (which only happens after `Hijack`) until we're sure it's a `101 Switching Protocols`. I'm storing the status and headers mostly for testing.

Reply:
I'm pretty sure we have to check the selected protocol is valid, but I'll verify